include/nfl/gmp.hpp

```diff
@@ -0,0 +1,220 @@
+#ifndef NFL_GMP_HPP
+#define NFL_GMP_HPP
+
+#include <vector>
+#include "nfl/poly.hpp"
+
+namespace nfl {
+
+template <class T, size_t Degree, size_t NbModuli>
+typename poly<T, Degree, NbModuli>::GMP poly<T, Degree, NbModuli>::gmp;
+
+/*
+ * Constructors
+ */
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(mpz_t const& v) {
+  set_mpz(v);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(mpz_class const& v) {
+  set_mpz({v});
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(std::initializer_list<mpz_class> const& values) {
+  set_mpz(values);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(std::array<mpz_t, Degree> const& values) {
+  set_mpz(values);
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+poly<T, Degree, NbModuli>::poly(std::array<mpz_class, Degree> const& values) {
+  set_mpz(values);
+}
+
+/*
+ * set_mpz's functions
+ */
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(mpz_t const& v) {
+  set_mpz({mpz_class(v)});
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(mpz_class const& v) {
+  set_mpz({v});
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(std::array<mpz_t, Degree> const &values) {
+  set_mpz(values.begin(), values.end());
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(std::array<mpz_class, Degree> const& values) {
+  set_mpz(values.begin(), values.end());
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+void poly<T, Degree, NbModuli>::set_mpz(
+    std::initializer_list<mpz_class> const& values) {
+  set_mpz(values.begin(), values.end());
+}
+
+template <class T, size_t Degree, size_t NbModuli>
+template <class It>
+void poly<T, Degree, NbModuli>::set_mpz(It first, It last) {
+  // CRITICAL: the object must be 32-bytes aligned to avoid vectorization issues
+  assert((unsigned long)(this->_data) % 32 == 0);
+
+  auto* iter = begin();
+  auto viter = first;
+
+  size_t size = std::distance(first, last);
+  // If the initializer has no more values than the polynomial degree use them
```

Comment:
the code would be much cleaner with the rror check right now, something like
 ``̀
if(size >degree) throw...
// regular path
for(...)